src/audio/rsp_mixer.S

```diff
@@ -0,0 +1,1133 @@
+
+	# Update the output pointer in RDRAM for next loop.
+	sll t0, k1, 2
+	lw s0, %lo(OUTPUT_RDRAM)
+	add s1, s0, t0
+	sw s1, %lo(OUTPUT_RDRAM)
+
+	# DMA the output buffer into RDRAM (s0 is the output pointer before update).
+	# We can do the transfer in background because it will surely be finished
+	# by the time we start filling the output area again.
+	addi t0, -1
+	jal DMAOutAsync
+	li s4, %lo(OUTPUT_AREA)
+
+	j MainLoop
+	nop
+
+End:
+	# Store the current channel volume in DMEM (permanent state)
+	jal EndMixer
+	nop
+
+#if VOLUME_FILTER
+	jal DMAFilterState
+	li t2, DMA_OUT_ASYNC
+#endif
+
+	# Wait for the last out transfer to be finished
+	jal DMAWaitIdle
+	nop
+
+	# Bye bye!
+	break
+
+
+###############################################################
+# DMAFilterState - Load/save the volume filter state via DMA.
+#
+# Arguments:
+#   t2:  DMA_* flag for DMAExec
+###############################################################
+
+	.func DMAFilterState
+DMAFilterState:
+	# Save volume filter state
+	lw s0, %lo(STATE_RDRAM)
+	li s4, %lo(XVOL_L)
+	j DMAExec
+	li t0, DMA_SIZE(MAX_CHANNELS*2*2, 1)
+	.endfunc
+
+
+
+###############################################################
+# UpdateAndFetch - Resampling loop.
+#
+# This function goes through all active channels, fetch
+# the samples of current waveforms via DMA, and apply the
+# required resampling (frequency change) for a specified
+# number of ticks (output samples).
+#
+# All channels parameters are fetched from WAVEFORM_SETTINGS.
+#
+# Arguments:
+#
+#  k0:  number of active channels
+#  k1:  number of ticks to resample
+#
+###############################################################
+
+	#define ticks          t3
+	#define wv_pos         t4
+	#define wv_step        t5
+	#define wv_len         t6
+	#define wv_loop_len    t7
+	#define wv_addr        t8
+	#define wv_dma_addr    v0
+	#define nchan          v1
+	#define dma_cache_end  s7
+	#define out_ptr        s5
+	#define wv_pos_to_dmem s6
+	#define wv_step_8x     t2
+	#define is_stereo      fp
+
+	.func UpdateAndFetch
+UpdateAndFetch:
+	move ra2, ra
+
+	li s5, %lo(CHANNEL_BUFFER)
+	li t0, (MAX_SAMPLES_PER_LOOP * MAX_CHANNELS * 2) / 64 - 1
+ClearLoop:
+	sqv v_zero,0, 0,s5
+	sqv v_zero,0, 1,s5
+	sqv v_zero,0, 2,s5
+	sqv v_zero,0, 3,s5
+	addi s5, 64
+	bnez t0, ClearLoop
+	addi t0, -1
+
+	li s1, %lo(WAVEFORM_SETTINGS)
+	li nchan, 0
+
+	li dma_cache_end, %lo(DMEM_SAMPLE_CACHE+SAMPLE_CACHE_SIZE)
+	sll dma_cache_end, WAVERFORM_POS_FRAC_BITS
+
+	# main update loop: will be done once per channel.
+UpdateLoop:
+	# Fetch waveform parameters. Notice that if the RDRAM
+	# address is zero, no waveform was configured in this channel,
+	# so we can skip directly to next channel.
+	lw wv_addr,     16(s1)
+	beqz wv_addr, WaveLoopEpilog2
+	lw wv_pos,       0(s1)
+	lw wv_step,      4(s1)
+	lw wv_len,       8(s1)
+	lw wv_loop_len, 12(s1)
+
+	# Prepare output pointer
+	addi out_ptr, nchan, %lo(CHANNEL_BUFFER)
+	add out_ptr, nchan
+	move ticks, k1
```

Comment:
Done. But I can't do that for function arguments in general because they're function-relative, so I can't easily add defines to the calling sites.

I think maybe the DMA function are the one creating more issues here. I think those might have to be learnt by heart but in general s0/s4 are often input/output, and t0 is count. t2=pitch (not used in this microcode) is possibly the odd one. If you see Mixer, it also takes s4 as its output argument.

Reply:
That's fair. Also see my previous comment.